gallery_dl/extractor/acqq.py

```diff
@@ -0,0 +1,152 @@
+            "artist"  : data['artist']['nick'],
+        }
+
+    @staticmethod
+    def images(page):
+        data = AcqqChapterExtractor._extract_data(page)
+        return [
+            (picture['url'], None)
+            for picture in data['picture']
+        ]
+
+    @staticmethod
+    def _extract_data(page):
+        data_enc = text.extract(page, "DATA = '", "'")[0]
+        nonce_expr = re.findall(AcqqChapterExtractor.nonce_decl_pattern, page)[0]
+        nonce = AcqqChapterExtractor._eval_js_expr(nonce_expr)
+        return json.loads(AcqqChapterExtractor._decode_chapter_data(data_enc, nonce))
+
+    nonce_decl_pattern = r"window.*n.*o.*n.*c.*e.*= ([^'].*);"
+    @staticmethod
+    def _decrypt_with_nonce(blob, nonce):
+      """Decrypt a blob of data given the correct nonce"""
+      arr = list(blob.strip())
+      mtc = re.findall(r'(\d+[a-zA-Z]+)', nonce)
+      mtc.reverse()
+      for v in mtc:
+        off = int(re.match(r'\d+', v).group(0)) & 0xFF
+        txt = re.sub(r'\d+', '', v)
+        arr[off:off+len(txt)] = []
+      return ''.join(arr) + '==='
+
+    @staticmethod
+    def _decode_chapter_data(data_enc, nonce):
+      """Decode chapter data given the base64-encoded data and the nonce"""
+      data = base64.b64decode(AcqqChapterExtractor._decrypt_with_nonce(data_enc, nonce))
+      data = data.decode('utf-8', errors='ignore')
+      data = re.sub(r'[\x00-\x1f\x7f-\x9f]', '', data)
+      return data
+
+    # The worst javascript "runtime" you've ever seen
+    @staticmethod
+    def _eval_js_expr(js_expr):
+      """Evaluate a subset of JavaScript expressions needed to calculate the decryption nonce"""
+      js_expr = re.sub(r"\((.*?)\).toString\(\)", r"str(\1)", js_expr)
+      js_expr = re.sub(r"eval\(\"(.*?)\"\)", r"\1", js_expr)
+      js_expr = re.sub(r"Math.pow\((\d+),(\d+)\)", r"(\1**\2)", js_expr)
+      js_expr = re.sub(r"(document\.(children|getElements?ByTagName\('html'\))|window\.Array)", "1", js_expr)
+      js_expr = re.sub(r"!!!", r"!", js_expr)
+      js_expr = re.sub(r"!!(\d+)", r"int(not (not \1))", js_expr)
+      js_expr = re.sub(r"\~([0-9.]+)", r"~int(\1)", js_expr)
+      js_expr = re.sub(r"!(\d+)", r"int(not \1)", js_expr)
+      js_expr = re.sub(r"parseInt", "int", js_expr)
+      js_expr = re.sub(r"\.substring\((\d+)\)", r"[\1:]", js_expr)
+      js_expr = re.sub(r"\+(['\"])", r"\1", js_expr)
+      js_expr = re.sub(r"(if|while|for)", "", js_expr)
+      js_expr = re.sub(r"Math.round", "js_round", js_expr)
+      js_expr = re.sub(r"'(.)'.charCodeAt\(\)", r"ord('\1')", js_expr)
+      js_expr = re.sub(r"(\d+)(<|<=|==|!=|=>|>)(\d+)\?(\d+):(\d+)", r"\4 if \1 \2 \3 else \5", js_expr)
+      js_expr = re.sub(r"([^0-9])\.([^0-9])", r"\1\2", js_expr)
+      py_expr = js_expr
+
+      # Now we run the translated code in a safe sandbox
+      # We'll also add a few functions needed by the translated code
+      result = eval(py_expr, {'__builtins__': {
```

Comment:
Thanks for the PR, but this right here is why I will definitely **not** merge it as it currently is. You must never `eval()` untrusted code from external sources.

The nonce expression seems easy enough to reverse engineer without invoking `eval()`. Just a string, followed by some sort of numeric expression, followed by another string etc.
```
        window["no"+"nce"] = "f0d7a" + (+eval("parseInt(7/3)")).toString() + "28b20d88" + (+eval("!!1+!1+!!2+!!2+1")).toString() + "7da3b8fa350" + (+eval("1 * (!window.Array) + 1")).toString() + "1dd" + (+eval("!!1*5")).toString() + "1";
        window["non"+"ce"] = "0db505bd0f" + (+eval("2")).toString() + "c3ab9f6560" + (+eval("0<=3?2:1")).toString() + (+eval("parseInt(7/3)")).toString() + "d36efa6d2";
```

youtube-dl/yt-dlp have a [jsinterp.py](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/jsinterp.py) module that does not use `eval` and we might be able to "borrow" from if nothing else helps.